File: Assets/Script/GamePlay/Buff/BuffPreciseShot.cs
using UnityEngine;
using GamePlay.Units;
using Status.damage;

namespace GamePlay.Buff
{
    public class BuffPreciseShot : BuffBase
    {
        [SerializeField] private int minDistance = 2;
        [SerializeField] private int bonusDamage = 2;

        public BuffPreciseShot()
        {
            MaxStacks = 1;
            DecayAtTurnStart = false;
        }

        public override void OnApply(MapUnit owner)
        {
            base.OnApply(owner);
            Initialize(owner);
        }

        public override void OnOutgoingDamage(ref float damage, DamageInfo damageInfo)
        {
            if (Owner == null || damageInfo.targetUnit == null) return;

            int dist = Mathf.Abs(Owner.gridPosition.x - damageInfo.targetUnit.gridPosition.x)
                     + Mathf.Abs(Owner.gridPosition.z - damageInfo.targetUnit.gridPosition.z);

            if (dist > minDistance)
            {
                damage += Stacks * bonusDamage;
            }
        }
    }
}

File: Assets/Lua/Buffs/BuffPreciseShot.lua
local BuffBase = require("Buffs.BuffBase")
local log = require("Logger")
---@class BuffPreciseShot : BuffBase
---@diagnostic disable-next-line: undefined-field
local BuffPreciseShot = BuffBase.extends(BuffBase)

function BuffPreciseShot:ctor(stacks)
    self.super.ctor(self, stacks or 1)
    self.MaxStacks = 1
    self.DecayAtTurnStart = false
    self.MinDistance = 2
    self.BonusDamage = 2
end

function BuffPreciseShot:OnApply(owner) end

function BuffPreciseShot:OnTurnStart(owner) end

function BuffPreciseShot:OnOutgoingDamage(damage, info)
    if self._Owner and info.targetUnit then
        local dx = self._Owner.gridPosition.x - info.targetUnit.gridPosition.x
        local dz = self._Owner.gridPosition.z - info.targetUnit.gridPosition.z
        local dist = math.abs(dx) + math.abs(dz)
        if dist > self.MinDistance then
            damage = tonumber(damage) + self.Stacks * self.BonusDamage
        end
    end
    log.logd("BuffPreciseShot:OnOutgoingDamage: %s", tostring(damage))
    return damage
end

return BuffPreciseShot
